File: 실습/4장/07_object_companion/Main.java
public class Main {

    static class Logger {
        private static final Logger INSTANCE = new Logger();

        private Logger() {
        }

        public static Logger getInstance() {
            return INSTANCE;
        }

        public void log(String message) {
            System.out.println("[LOG] " + message);
        }
    }

    static class User {
        private final String name;

        private User(String name) {
            this.name = name;
        }

        public static User create(String name) {
            return new User(name);
        }
    }

    public static void main(String[] args) {
        Logger.getInstance().log("프로그램 시작");

        User user = User.create("철수");
        System.out.println(user.name);
    }
}

File: 실습/4장/07_object_companion/Main.kt
object Logger {
    fun log(message: String) {
        println("[LOG] $message")
    }
}

class User private constructor(
    val name: String
) {
    companion object {
        fun create(name: String): User {
            return User(name)
        }
    }
}

fun main() {
    Logger.log("프로그램 시작")

    val user = User.create("철수")
    println(user.name)
}